src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -3,26 +3,101 @@
 
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
+from .candidates import is_already_installed
+
 if MYPY_CHECK_RUNNING:
-    from typing import Any, Dict, Optional, Sequence, Tuple, Union
+    from typing import Any, Dict, Optional, Sequence, Set, Tuple, Union
 
     from pip._internal.req.req_install import InstallRequirement
+    from pip._vendor.packaging.version import _BaseVersion
 
     from .base import Requirement, Candidate
     from .factory import Factory
 
+# Notes on the relationship between the provider, the factory, and the
+# candidate and requirement classes.
+#
+# The provider is a direct implementation of the resolvelib class. Its role
+# is to deliver the API that resolvelib expects.
+#
+# Rather than work with completely abstract "requirement" and "candidate"
+# concepts as resolvelib does, pip has concrete classes implementing these two
+# ideas. The API of Requirement and Candidate objects are defined in the base
+# classes, but essentially map fairly directly to the equivalent provider
+# methods. In particular, `find_matches` and `is_satisfied_by` are
+# requirement methods, and `get_dependencies` is a candidate method.
+#
+# The factory is the interface to pip's internal mechanisms. It is stateless,
+# and is created by the resolver and held as a property of the provider. It is
+# responsible for creating Requirement and Candidate objects, and provides
+# services to those objects (access to pip's finder and preparer).
+
 
 class PipProvider(AbstractProvider):
     def __init__(
         self,
         factory,  # type: Factory
         constraints,  # type: Dict[str, SpecifierSet]
         ignore_dependencies,  # type: bool
+        upgrade_strategy,  # type: str
+        roots,  # type: Set[str]
     ):
         # type: (...) -> None
         self._factory = factory
         self._constraints = constraints
         self._ignore_dependencies = ignore_dependencies
+        self._upgrade_strategy = upgrade_strategy
+        self.roots = roots
```

Comment:
I wonder if this can be named more specific, e.g. `root_requirement_names`?

Reply:
I feel like `user_provided_requirement_names` is MUCH better, since it makes it very clear what this is for. I'm biased toward *slightly* longer names when they convey enough detail.

Reply:
I changed it from `root_reqs` because we're massively over-using "req" and these are names, not requirements, anyway.

I'm not a huge fan of over-long names (they tend to mean you end up needing to break lines, and then I just have arguments with linters on how to do the breaks) but this isn't used very often, so I'll look at what I can do. I do prefer "user_provided" over "root" slightly - I feel like "root" would be better if it was sufficiently precise, but I'm not sure it is...

I'm waffling now, I'll go with a longer name based on "user provided".

Reply:
I went with `user_requested`. Because it's a keyword parameter, making it *too* long makes the it very verbose when calling the constructor.